src/util/loop_profiler.hpp

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include <chrono>
+#include <numeric>
+#include <optional>
+#include <thread>
+#include <unordered_map>
+
+#include <ros/console.h>
+
+/**
+ * @brief Profiles the execution time of a loop composed of multiple events.
+ */
+class LoopProfiler {
+private:
+    using Clock = std::chrono::high_resolution_clock;
+    using EventReadings = std::vector<Clock::duration>;
+    using Readout = std::chrono::milliseconds;
+
+    std::string mName;
+    size_t mPrintTick;
+    std::unordered_map<std::string, EventReadings> mEventReadings;
+
+    std::optional<Clock::time_point> mLastEpochTime;
+    size_t mTick = 0; // Loop iteration counter
+
+public:
+    LoopProfiler(std::string_view name, size_t printTick = 60) : mName{name}, mPrintTick{printTick} {}
+
+    /**
+     * @brief Call this at the beginning of each loop iteration.
+     */
+    void beginLoop() {
```

Comment:
can we add a param or something to suppress this profiler output? since its printing to ros info we cant just turn off output=screen right?

Reply:
I can make it print less, I still think the info is useful. Worst case we can turn off output=screen yeah

Reply:
since its rosinfo it bypasses output=screen

Reply:
I see, I will make a ros param for this